src/Confluent.Kafka/AdminClient.cs

```diff
@@ -0,0 +1,750 @@
+        /// </summary>
+        public Metadata GetMetadata(bool allTopics, string topic, TimeSpan timeout)
+            => kafkaHandle.GetMetadata(
+                allTopics, 
+                topic == null ? null : getKafkaTopicHandle(topic), 
+                timeout.TotalMillisecondsAsInt());
+
+
+        /// <summary>
+        ///     Refer to <see cref="Confluent.Kafka.IClient.AddBrokers(string)" />
+        /// </summary>
+        public int AddBrokers(string brokers)
+            => kafkaHandle.AddBrokers(brokers);
+
+
+        /// <summary>
+        ///     Refer to <see cref="Confluent.Kafka.IClient.Name" />
+        /// </summary>
+        public string Name
+            => kafkaHandle.Name;
+
+
+        /// <summary>
+        ///     An opaque reference to the underlying librdkafka 
+        ///     client instance.
+        /// </summary>
+        public Handle Handle
+            => handle;
+
+
+        /// <summary>
+        ///     Releases all resources used by this AdminClient. In the current
+        ///     implementation, this method may block for up to 100ms. This 
+        ///     will be replaced with a non-blocking version in the future.
+        /// </summary>
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+
+        /// <summary>
+        ///     Releases the unmanaged resources used by the
+        ///     <see cref="Confluent.Kafka.AdminClient" />
+        ///     and optionally disposes the managed resources.
+        /// </summary>
+        /// <param name="disposing">
+        ///     true to release both managed and unmanaged resources;
+        ///     false to release only unmanaged resources.
+        /// </param>
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                callbackCts.Cancel();
+                try
+                {
+                    callbackTask.Wait();
+                }
+                catch (AggregateException e)
+                {
+                    if (e.InnerException.GetType() != typeof(TaskCanceledException))
+                    {
+                        throw e.InnerException;
+                    }
+                }
+                finally
+                {
+                    callbackCts.Dispose();
+                }
+
+                DisposeResources();
+            }
+        }
+
+
+        private void DisposeResources()
+        {
+            kafkaHandle.DestroyQueue(resultQueue);
+
+            foreach (var kv in topicHandles)
+            {
+                kv.Value.Dispose();
+            }
+
+            if (handle.Owner == this)
+            {
+                ownedClient.Dispose();
+            }
+        }
+
+
+        /// <summary>
+        ///     Refer to <see cref="Confluent.Kafka.IClient.OnStatistics" />
+        /// </summary>
+        public event EventHandler<string> OnStatistics
```

Comment:
These don't belong on the constructor for the same reason we don't have individual arguments on the `ProducerAsync` methods. Note that I think it's ok to make an exception for serializers / deserializers because they directly correspond to the generic type arguments (it's quite nice).

These handlers belong in the config dictionary. I've already moved `OnLog` there, since it's the most important, but as you point out, we want to move `OnError` and `OnStatistics` as well.

The disadvantage of course is lack of static type checking. But we already threw that out the window with the other config properties - this is no different. One could argue we the `IEnumerable<KeyValuePair<string, object>>` config interface is a bad idea, but actually, I think this works out quite well because there is a need for an intermediary interface, and this does the job nicely. Ultimately, we probably want a `Config` class that allows strongly typed configuration. We probably also want a class that understands standard ASP.NET configuration. Both can build on `IEnumerable<KeyValuePair<string, object>>`.

In the future, we can add functionality such that if the handler is configured as a string, we could use reflection to search for the class and use that. If configured as an object, use the object directly. But not yet.

An alternative to all the above that would allow us to continue to use .NET events for these handlers is to cache results that occur before a handler is registered (and emit them when the handler is known), but that's not as good structurally, and would need some logic to guard against the case where no user defined handler is added... a bit messy.